modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,701 @@
+private proc _expBySquaring(x: ?t, n): t {
+    // TODO -- _expBySquaring(pinv(x), -n);
+    if n < 0  then halt("Negative powers not yet supported");
+    else if n == 0  then return eye(x.domain);
+    else if n == 1  then return x;
+    else if n%2 == 0  then return _expBySquaring(dot(x, x), n / 2);
+    else return dot(x, _expBySquaring(dot(x, x), (n - 1) / 2));
+}
+
+
+/* Return element-wise power: ``a**B`` */
+proc matPow(a, B: [?Bdom] ?eltType) where isNumeric(a) {
+  var C: [Bdom] eltType = a**B;
+  return C;
+}
+
+
+/* Return cross-product of 3-element vectors ``A`` and ``B`` with domain of
+  ``A`` */
+proc cross(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+  if Adom.size != 3 || Bdom.size != 3 then
+    halt("cross() expects arrays of 3 elements");
+
+  var C = Vector(Adom, eltType);
+
+  // TODO -- a better way to do 0-based indexing...
+  C[_raw(Adom, 0)] = A[_raw(Adom, 1)]*B[_raw(Bdom, 2)] -
+                     A[_raw(Adom, 2)]*B[_raw(Bdom, 1)];
+
+  C[_raw(Adom, 1)] = A[_raw(Adom, 2)]*B[_raw(Bdom, 0)] -
+                     A[_raw(Adom, 0)]*B[_raw(Bdom, 2)];
+
+  C[_raw(Adom, 2)] = A[_raw(Adom, 0)]*B[_raw(Bdom, 1)] -
+                     A[_raw(Adom, 1)]*B[_raw(Bdom, 0)];
+
+  return C;
+}
+
+
+pragma "no doc"
+/* Compute 0-based index */
+private inline proc _raw(D: domain(1), i) {
+  return D.dim(1).orderToIndex(i);
+}
+
+
+//
+// Matrix Structure
+//
+
+/* Return lower triangular part of matrix, below the diagonal + ``k`` */
+proc tril(A: [?D] ?eltType, k=0) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  var L = Matrix(A);
+  forall (i, j) in D do
+    if (i < j-k) then L[i, j] = 0: eltType;
+  return L;
+}
+
+
+/* Return upper triangular part of matrix, above the diagonal + ``k`` */
+proc triu(A: [?D] ?eltType, k=0) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  var U = Matrix(A);
+  const zero = 0: eltType;
+  forall (i, j) in D do
+    if (i > j-k) then U[i, j] = zero;
+  return U;
+}
+
+
+/* Return `true` if matrix is diagonal */
+proc isDiag(A: [?D] ?eltType) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+
+  const zero = 0: eltType;
+  for (i, j) in D {
+    if i != j && A[i, j] != zero then return false;
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is Hermitian */
+proc isHermitian(A: [?D]) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  for (i, j) in D {
+    if i > j {
+      if A[i, j] != conjg(A[j, i]) then return false;
+    }
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is symmetric */
+proc isSymmetric(A: [?D]) : bool {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  for (i, j) in D {
+    if i > j {
+      if A[i, j] != A[j, i] then return false;
+    }
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is lower triangular below the diagonal + ``k`` */
```

Comment:
I would just cut and paste the additional documentation from ``tril`` and ``triu`` here (explaining ``k`` explicitly), or link these to those routines...